torch/_inductor/kernel/flex_attention.py

```diff
@@ -342,77 +342,163 @@ def forward_inner(
+                    # accumulated values
+                    acc, l_i, m_i,
+                    # Offsets
+                    off_z, off_h, offs_m, offs_n,
+                    MATMUL_PRECISION, RCP_LN2,
+                    {{gen_argdefs()}},
+                    IS_FULL_BLOCKS,
+                )
+
+                # update pointers
+                offset = get_offset_for_next_block(
+                    start_n, kv_indices, kv_num_blocks,
+                    SPARSE_KV_BLOCK_SIZE, SPARSE_KV_MULTIPLE, BLOCK_N
+                )
+
+                V_block_ptr = tl.advance(V_block_ptr, (offset, 0))
+                K_block_ptr = tl.advance(K_block_ptr, (0, offset))
+
+                offs_n = offs_n + offset
+
+            # Need to mask out part of the lask block for non divisible seqlen.
+            acc, l_i, m_i = fwd_compute_block_mn(
+                q, K_block_ptr, V_block_ptr, Q_LEN, KV_LEN,
+                # accumulated values
+                acc, l_i, m_i,
+                # Offsets
+                off_z, off_h, offs_m, offs_n,
+                MATMUL_PRECISION, RCP_LN2,
+                {{gen_argdefs()}},
+                IS_FULL_BLOCKS, True,
+            )
+
+
+    else:
+        # loop over k, v and update accumulator until block_n_end
+        for start_n in range(block_n_start, block_n_end):
+            acc, l_i, m_i = fwd_compute_block_mn(
+                q, K_block_ptr, V_block_ptr, Q_LEN, KV_LEN,
+                # accumulated values
+                acc, l_i, m_i,
+                # Offsets
+                off_z, off_h, offs_m, offs_n,
+                MATMUL_PRECISION, RCP_LN2,
+                {{gen_argdefs()}},
+                IS_FULL_BLOCKS,
+            )
+
+            # update pointers
+            offset = get_offset_for_next_block(
+                start_n, kv_indices, kv_num_blocks,
+                SPARSE_KV_BLOCK_SIZE, SPARSE_KV_MULTIPLE, BLOCK_N
+            )
+
+            V_block_ptr = tl.advance(V_block_ptr, (offset, 0))
+            K_block_ptr = tl.advance(K_block_ptr, (0, offset))
+
+            offs_n = offs_n + offset
+
+    return acc, l_i, m_i
+
+"""
+
+
+fwd_compute_block_mn = r"""
+@triton.jit
+def fwd_compute_block_mn(
+    q, K_block_ptr, V_block_ptr, Q_LEN, KV_LEN,
+    # accumulated values
+    acc, l_i, m_i,
+    # Offsets
+    off_z, off_h, offs_m, offs_n,
+    MATMUL_PRECISION, RCP_LN2,
+    {{gen_argdefs()}},
+    IS_FULL_BLOCKS, is_last_block=False,
+):
+    # Redefines all kernel parameters (BLOCK_M, etc.) so we don't need to plumb them all through
+    {{gen_defines() | indent_except_first(1)}}
+
+    # -- load k --
+    k = tl.load(K_block_ptr, boundary_check=(1,))
+    # -- compute qk ---
+    qk = tl.dot(q, k) # TODO: use cuda matmul when q_len <= 2.
+    if not PRESCALE_QK:
+        qk *= SM_SCALE
+    # ~~~~~~~~~~~~~~~~~~~ Apply score modification  ~~~~~~~~~~~~~~~~~~~
+    if is_last_block:
+        # Applying mod to index for last block for non divisible seqlen.
```

Comment:
For my understanding, we are essentially, reloading and applying the mod on the [N: ceildiv(n/block-size)*blocksize] elements of the sequence. This allows us to worry about any out of bound shenanigans at the cost of some wasted work. 

we then apply a mask to the post_mod scores to ignore these elements. 

That is cool and I like it but maybe we leave a commet to that affect since it took me a second to figure out why the mod was there